Feature/workflow support

987Nabil opened the pull request:
Impl. #7

987Nabil commented:
@djspiewak 
I do not want flying to mars and then realising we just want to go to the moon.
So could you take a look and tell me, if I am in the right space ship? 🚀  😃

djspiewak commented:
This is a really nice start! Definitely headed in the right direction, though I have some aesthetic quibbles with file organization. :-P

987Nabil commented:
Still wip ;) 

I realised until now, you have the data model separated from the rendering logic. Would you like to keep this?
Or you think my mixed approach is ok too?

As you can see, I copied some useful methods from `GenerativePlugin`. Do you have any suggestion where to put such code for usage in both files?

My gut feeling says, that `GenerativePlugin` is quite big.
Is working on this topic maybe a good place to rethink the current design and maybe split the file a bit? 🤔

987Nabil commented:
I plan not to implement `branches-ignore` or `tags-ignore` for events that would support it.
Since one can not mix it with `branches` and `tags`, but there is the negation syntax `!<branchPattern>`.

> You can exclude tags and branches using the ! character. The order that you define patterns matters.
> 
> A matching negative pattern (prefixed with !) after a positive match will exclude the Git ref.
> A matching positive pattern after a negative match will include the Git ref again.

I'll add a part about it in the docs.
See https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-using-positive-and-negative-patterns

987Nabil commented:
@djspiewak I think the model should fit now, as well as the generation.
I would add docu after you checked the code.
I tried to go with a minimal code change approach. I also don't think this changes should be breaking in any means, since the usage of `Workflow` in the old logic is not reachable from the public, if I see this correctly.

Right now none of the settings of the generate plugin are applying to custom workflows.
I think this provides the most flexibility.

Imo the current state should be fine for an initial release of the feature.

987Nabil commented:
Open questions:
- `githubWorkflow` settings like the Scala version, OSs or the definition for runs-on are the same for all jobs. Should that be customisable?
- Should there be more validation? Cron expressions for the `Schedule` trigger event for example?

987Nabil commented:
@djspiewak I am using this version currently in multiple projects and it works well so far. I changed to the way `oses` are compiled and added the `GenerationTarget` since I realised I don't want to have CI and Custom workflows.
But I think the current solution should be flexible enough for all needs.

catostrophe commented:
@987Nabil does this support adding the `workflow_dispatch` event as a trigger?

987Nabil commented:
@catostrophe For custom workflows yes. For the default CI job aka `ci.yml` no. `push` and `pull_request` are currently still hard coded for the ci job.
It is in this state, since the intention is not to break any old code.

catostrophe commented:
@987Nabil that's exactly what I need - to tweak the default ci.yml to run on `workflow_dispatch` event. I wanted to add this as a small PR via a flag

@djspiewak what's the best way to introduce this feature? as a separate small PR. as a part of this one?

987Nabil commented:
@catostrophe my suggestion would be: Merge this PR first, then do it separately afterward. I think this PR is quite big already. Even though I guess it might be a small change.

987Nabil commented:
@djspiewak I just merged the main. And I am already heavily using it with a local build. Would be nice if you could review this and I could use an official build :)

djspiewak commented:
@987Nabil Sorry for the delay! Looking at this now

987Nabil commented:
I just saw the one test was failing. I fixed it. Was a bug in the test :/

987Nabil commented:
@djspiewak I updated the PR with support for reusable workflows and merged in the master.
If you need any kind of help, let me know. I would really love to merge this soonish 😉

catostrophe commented:
Any chance to resurrect this PR?

987Nabil commented:
@eed3si9n I am currently using a self build version that is based on this pr and since nothing happens for so long I was going to create a new lib. Which I actually don't like, because i think one lib of it's kind is more than enough.
But since this is living now in the SBT org, I thought I ask you, if you are fine with me reviving this pr? If so, it would be great if you can comment on the general structure of the or, so i could then go on to fix/update it :slightly_smiling_face:

robmwalsh commented:
It would be great to have this revived/merged!

eed3si9n commented:
@987Nabil 

> But since this is living now in the SBT org, I thought I ask you, if you are fine with me reviving this pr? If so, it would be great if you can comment on the general structure of the or, so i could then go on to fix/update it 🙂

I am here as a backstop, but generally the maintainers are @sbt/sbt-github-actions team.
If there are no breaking changes to the existing builds, I think the PR looks pretty good already. Maybe it should be better documented in README?

mdedetrich commented:
I can have a look at it, but to get the premise of the PR by custom workflow you mean the ability to generate a completely custom github action workflow flow with the various actions.

If so I can try to look at it during in a week (I have a conference to go to). In the meantime if you could rebase the PR to fix the current merge conflicts that would be great!

987Nabil commented:
@eed3si9n sorry, I just saw SBT org and you changing something, so I assumed 😅

@mdedetrich basically it aims to be a scala DSL for actions/workflows. And yes, the goal is to not change any api but just extend it.

mdedetrich commented:
@987Nabil Thanks for the response, as mentioned before I will realistically look at it from next week Monday but if I can find some time I might start earlier.

I suspect that the reason behind the merge conflicts is the permissions feature at https://github.com/sbt/sbt-github-actions/pull/105 which I wrote and was consequently merged. On the surface I think it actually makes sense to incorporate this into your DSL when doing the rebase.

987Nabil commented:
@mdedetrich there where also changes to the GitHub api. So I'd like to update anyway. I'll check on it today or maybe tomorrow.

mdedetrich commented:
Perfect thanks!

987Nabil commented:
@mdedetrich Sorry it might take longer than I expected. Maybe I can find some time today.

mdedetrich commented:
No worries, no rush

987Nabil commented:
@mdedetrich I guess this is now in a reviewable state. The branch is up to date and I migrated the permissions. I simplified and generalized some code a bit. This comes with a little less beautiful workflows (see the missing empty lines), but is an overall benefit I guess, also in terms of maintainability.
I am happy to add more docu, if you can give me a hint what you would expect.
Should I add an example?

Also I did not check, if this state on the same feature level as the latest github api.
But I guess it is ok to add stuff later if missing. The PR is already large.

mdedetrich commented:
@987Nabil Thanks, I will start looking at it in a few days. Note that some of the scripted tests are still failing

mdedetrich commented on src/main/scala/sbtghactions/GenerativePlugin.scala:
```diff
@@ -617,16 +617,6 @@ ${indentOnce(workflow.jobs.map(compileJob(_, sbt)).mkString("\n\n"))}
     } else {
       githubWorkflowSbtCommand.value
     }
-    //compileWorkflow(
```
@987Nabil What is the context behind this test being commented, i.e. is it temporary to pass CI or is it intentional?

987Nabil replied:
It was only an example I copied for how the current call looks like. I forgot to remove it.

987Nabil replied:
I just realised it is already gone.

mdedetrich commented:
Reviewing this today

mdedetrich commented on src/main/scala/sbtghactions/LabelEventType.scala:
```diff
@@ -0,0 +1,25 @@
+ */
+
+package sbtghactions
+
```
Can we add documentation with the `@see` to the equivalent github documentation (incase something changes/breaks), see https://github.com/sbt/sbt-github-actions/blob/main/src/main/scala/sbtghactions/PermissionScope.scala#L21-L23 for an example

mdedetrich commented on src/main/scala/sbtghactions/MilestoneEventType.scala:
```diff
@@ -0,0 +1,27 @@
+ */
+
+package sbtghactions
+
```
Can we add documentation with the `@see` to the equivalent github documentation (incase something changes/breaks), see https://github.com/sbt/sbt-github-actions/blob/main/src/main/scala/sbtghactions/PermissionScope.scala#L21-L23 for an example

mdedetrich commented on src/main/scala/sbtghactions/PREventType.scala:
```diff
@@ -16,23 +16,23 @@
 
 package sbtghactions
 
```
Can we add documentation with the `@see` to the equivalent github documentation (incase something changes/breaks), see https://github.com/sbt/sbt-github-actions/blob/main/src/main/scala/sbtghactions/PermissionScope.scala#L21-L23 for an example

mdedetrich commented on src/main/scala/sbtghactions/PRReviewCommentEventType.scala:
```diff
@@ -0,0 +1,25 @@
+ */
+
+package sbtghactions
+
```
Can we add documentation with the `@see` to the equivalent github documentation (incase something changes/breaks), see https://github.com/sbt/sbt-github-actions/blob/main/src/main/scala/sbtghactions/PermissionScope.scala#L21-L23 for an example

mdedetrich commented on src/main/scala/sbtghactions/PRReviewEventType.scala:
```diff
@@ -0,0 +1,25 @@
+ */
+
+package sbtghactions
+
```
Can we add documentation with the `@see` to the equivalent github documentation (incase something changes/breaks), see https://github.com/sbt/sbt-github-actions/blob/main/src/main/scala/sbtghactions/PermissionScope.scala#L21-L23 for an example

mdedetrich commented on src/main/scala/sbtghactions/PRTargetEventType.scala:
```diff
@@ -0,0 +1,36 @@
+ */
+
+package sbtghactions
+
```
Can we add documentation with the `@see` to the equivalent github documentation (incase something changes/breaks), see https://github.com/sbt/sbt-github-actions/blob/main/src/main/scala/sbtghactions/PermissionScope.scala#L21-L23 for an example

mdedetrich commented on src/main/scala/sbtghactions/PermissionScope.scala:
```diff
@@ -16,7 +16,30 @@
 package sbtghactions
 
-sealed trait Permissions extends Product with Serializable
+import sbtghactions.RenderFunctions.*
```
Can we change this to `._`

987Nabil replied:
I could. But why?
If you take a look in `sbt.Defaults`, from scala 2.12.13+ this is the prefered wildcard import. And I am using it in a lot of other places too, since intellij is reading the config and does his automatically.

```scala
    scalacOptions := {
      val old = scalacOptions.value
      if (sbtPlugin.value && VersionNumber(scalaVersion.value)
            .matchesSemVer(SemanticSelector("=2.12 >=2.12.13")))
        old ++ Seq("-Wconf:cat=unused-nowarn:s", "-Xsource:3")
      else old
    }
```

mdedetrich replied:
Its mainly to make sure the code is consistent, currently `.*` is not used anywhere else in the code so it sticks out as a sore thumb.

I am not against using `.*` in general however this is better tackled when adding scalafmt in this codebase, something I am planning to propose after this PR is merged.

987Nabil replied:
Hmm in that case, is it maybe okay to leave it as is, since it is used in many places in my PR, and save the effort with the goal in mind that scalafmt will ensure the consistency soon™?

mdedetrich replied:
Don't worry about it then, I will add a commit to the PR just before I merge it.

mdedetrich commented on src/main/scala/sbtghactions/ProjectCardEventType.scala:
```diff
@@ -0,0 +1,27 @@
+ */
+
+package sbtghactions
+
```
Can we add documentation with the `@see` to the equivalent github documentation (incase something changes/breaks), see https://github.com/sbt/sbt-github-actions/blob/main/src/main/scala/sbtghactions/PermissionScope.scala#L21-L23 for an example

mdedetrich commented on src/main/scala/sbtghactions/ProjectColumnEventType.scala:
```diff
@@ -0,0 +1,26 @@
+ */
+
+package sbtghactions
+
```
Can we add documentation with the `@see` to the equivalent github documentation (incase something changes/breaks), see https://github.com/sbt/sbt-github-actions/blob/main/src/main/scala/sbtghactions/PermissionScope.scala#L21-L23 for an example

mdedetrich commented on src/main/scala/sbtghactions/ProjectEventType.scala:
```diff
@@ -0,0 +1,28 @@
+ */
+
+package sbtghactions
+
```
Can we add documentation with the `@see` to the equivalent github documentation (incase something changes/breaks), see https://github.com/sbt/sbt-github-actions/blob/main/src/main/scala/sbtghactions/PermissionScope.scala#L21-L23 for an example

mdedetrich commented on src/main/scala/sbtghactions/RegistryPackageEventType.scala:
```diff
@@ -0,0 +1,24 @@
+ */
+
+package sbtghactions
+
```
Can we add documentation with the `@see` to the equivalent github documentation (incase something changes/breaks), see https://github.com/sbt/sbt-github-actions/blob/main/src/main/scala/sbtghactions/PermissionScope.scala#L21-L23 for an example

mdedetrich commented on src/main/scala/sbtghactions/ReleaseEventType.scala:
```diff
@@ -0,0 +1,29 @@
+ */
+
+package sbtghactions
+
```
Can we add documentation with the `@see` to the equivalent github documentation (incase something changes/breaks), see https://github.com/sbt/sbt-github-actions/blob/main/src/main/scala/sbtghactions/PermissionScope.scala#L21-L23 for an example

mdedetrich commented on src/main/scala/sbtghactions/WatchEventType.scala:
```diff
@@ -0,0 +1,23 @@
+ */
+
+package sbtghactions
+
```
Can we add documentation with the `@see` to the equivalent github documentation (incase something changes/breaks), see https://github.com/sbt/sbt-github-actions/blob/main/src/main/scala/sbtghactions/PermissionScope.scala#L21-L23 for an example

mdedetrich commented on src/main/scala/sbtghactions/WorkflowRunEventType.scala:
```diff
@@ -0,0 +1,24 @@
+ */
+
+package sbtghactions
+
```
Can we add documentation with the `@see` to the equivalent github documentation (incase something changes/breaks), see https://github.com/sbt/sbt-github-actions/blob/main/src/main/scala/sbtghactions/PermissionScope.scala#L21-L23 for an example

mdedetrich commented on README.md:
```diff
@@ -115,6 +116,7 @@ Any and all settings which affect the behavior of the generative plugin should b
 - `githubWorkflowJobSetup` : `Seq[WorkflowStep]` – The automatically-generated checkout, setup, and cache steps which are common to all jobs which touch the build (default: autogenerated)
 - `githubWorkflowEnv` : `Map[String, String]` – An environment which is global to the entire **ci.yml** workflow. Defaults to `Map("GITHUB_TOKEN" -> "${{ secrets.GITHUB_TOKEN }}")` since it's so commonly needed.
 - `githubWorkflowAddedJobs` : `Seq[WorkflowJob]` – A convenience mechanism for adding extra custom jobs to the **ci.yml** workflow (though you can also do this by modifying `githubWorkflowGeneratedCI`). Defaults to empty.
+- `githubWorkflowCustomWorkflows`: `Map[String, Workflow]` - This is the place to define your custom workflows. The key represents the filename (**.yml** is added if not provided) the value the workflow definition. The settings of the generative plugin apply, as long as they are not part of the workflow trigger events. For example, the `githubWorkflowTargetBranches` setting has no influence on cu ...
```
What about keys like `githubWorkflowJavaVersions` or `githubWorkflowOSes`? Is this supported (if not I suspect that this may be a case of feature creep but having the ability to interpolate these keys into custom workflows would be useful/beneficial if practical).

987Nabil replied:
I don't know what you have in mind, but this is now a very general API. It might be, that there is not even a java version needed for the workflow. Also GitHub has the feature of self hosted runners, that might not have the tags for the oses. So that might be confusing for the user. I guess using this keys should be up to the one defining the Workflows.

mdedetrich replied:
So to put it differently, what I had is having the ability to use the `githubWorkflowJavaVersions`/`githubWorkflowOSes`  values in a custom `ReusableWorkflowJob`. While its definitely true that it doesn't make sense in some cases, at least having access to those values can be useful for certain types of custom workflows which does PR implements.

What I had in mind is adjusting `ReusableWorkflowJob` so that it has the https://github.com/sbt/sbt-github-actions/pull/65/files#diff-535533c663b9848aff0c0b943149d3dd8b126b34aa463e507ada7e6011ce8288R38-R39 so that it has the `scalas` and `javas` variables.

Will add a comment there.

mdedetrich reviewed (changes requested):
Overall very nice work. I just have some minor comments (apologies for the spam regarding documentation for the DSL, its just easier to track).

I do have one question regarding how `githubWorkflowJavaVersions`/`githubWorkflowOSes`/ interact with this PR?

mdedetrich commented:
@987Nabil Did you have time to look into the review comments? Also @armanbilge incase you have time to review the PR yourself.

mdedetrich commented on src/main/scala/sbtghactions/WorkflowJob.scala:
```diff
@@ -16,6 +16,16 @@
+  name: String,
+  uses: WorkflowRef,
+  cond: Option[String] = None,
+  needs: List[String] = List(),
```
Should we also add `javas` and `scalas` here (like with `WorkflowJob`) so that these values can be accessable if someone wants to create a custom workflow that uses those values?

987Nabil replied:
I am not so sure about that. I think you can make a very good argument, that it is after this PR more or less a smell, that `WorkflowJob` does not just represent the GH api and that as a pure data structure should only do that.
And yes, I bring this argument, because I think it is the better design. What I did not think about yet is, if this would bing migration issues.
But I think, in a perfect world there would be no `oses`, `javas` and `scalas` here, but the plugin would build the right `WorkflowJob` out of the sbt keys before rendering it.

mdedetrich replied:
After thinking about it, I think it does make sense in a future PR because as you said some things might need to be moved around and this PR has already been open for some time. The reason I brought this up is because this is actually a usecase for one of the projects I maintain, i.e. we definitely have a use for custom workflows however we are still going to be running jobs on JVM's defined by `githubWorkflowJavaVersions` and OS's defined by `githubWorkflowOSes` (amongst other things) and without such a change you basically have to redefine the exact same values (once in `githubWorkflowJavaVersions` and another in your custom workflow) which also has the obvious risk of them accidentally going out of sync.

987Nabil replied:
You actually don't. You will define the workflow anyway in an sbt file or your own sbt plugin. And there you can know the key. You can just use it somewhat like this.

```scala
  val SetupScala: WorkflowStep.Use = WorkflowStep.Use(
    name = Some("Setup Java and Scala"),
    ref = UseRef.Public("coursier", "setup-action", "v1"),
    params = Map("jvm" -> GenerativeKeys.githubWorkflowJavaVersions.value.head.version),
    )
```

What I could understand is, having a method on the data types like "withJavaVersions" that sets a matrix up in the right way. But I would really remove every data from the structure that is not a reflection of the github api.

mdedetrich replied:
> If you like, we could try to have a call and try to talk about a strategy for this or pairing on a follow up issue.

Sure! Next week I will have some time (feel free to email me to discuss, its on my profile)

For now I will merge the PR in a few days to see if we can get any more additional reviews.

987Nabil commented on src/main/scala/sbtghactions/CheckRunEventType.scala:
```diff
@@ -0,0 +1,25 @@
+
+package sbtghactions
+
+sealed trait CheckRunEventType extends EventType
```
The super type `EventType` has already `Product with Serializable`. I am pretty sure that should be enough.

mdedetrich replied:
Apologies, I missed that `EventType` happens to already extend `Product with Serializable` so you can ignore this. Resolving.

987Nabil commented on src/main/scala/sbtghactions/GenerationTarget.scala:
```diff
@@ -0,0 +1,27 @@
+ */
+
+package sbtghactions
+
```
This is a plugin internal data structure and has nothing to do with GitHub.

mdedetrich commented on src/main/scala/sbtghactions/CheckSuiteEventType.scala:
```diff
@@ -0,0 +1,25 @@
+
+package sbtghactions
+
+sealed trait CheckSuiteEventType extends EventType
```
Apologies, I missed that `EventType` happens to already extend `Product with Serializable` so you can ignore this. Resolving.

mdedetrich reviewed (approved):
Approving PR

mdedetrich commented:
@armanbilge I will wait till mid next week if you have a few comments, at which point I will merge.

987Nabil commented:
@mdedetrich Is there something blocking this except the merge conflicts? If not I'd solve this asap to merge it.

mdedetrich commented:
@987Nabil 

> @mdedetrich Is there something blocking this except the merge conflicts? If not I'd solve this asap to merge it.

So I want to give some context behind whats going on, I spoke with @armanbilge and currently right now there is an issue where this project needs another active maintainer that actually uses this project (which is as of now is just me). Due to this I decided to try and make a `0.2.0` release with bug fixes/minor feature improvements that have been collected in this repository over time and have not been released yet. The reason why I want to get `0.2.0` out as fast as I can is because the current latest version of sbt-github-actions is referencing old versions of github actions (such as `setup-java`) that is generating warnings for projects (and there are other fixes as well).

Once `0.2.0` is released I want to properly look into this PR, which also means trying to use it in one of the projects that I maintain (likely one of the Pekko ones) to see if anything is missed since this feature is directly relevant and very useful for me specifically.

In regards to merge conflicts, there will be one more additional non trivial PR merge which is https://github.com/sbt/sbt-github-actions/pull/136. After that point I will upstream some of the relevant bug fixes in typelevels fork of sbt-github-actions (they ended up forking this project because it wasn't maintained) which may generate more conflicts.

If you have the time to merge the conflicts than that would be great, otherwise I can also just push the merge conflicts on your branch as the mentioned changes start rolling in.
